### PR TITLE
Feat: Add support for replaying :defined pseudo-class of custom elements

### PR DESCRIPTION
**Pull request by YunFeng0817**
Some websites use [:defined](https://developer.mozilla.org/en-US/docs/Web/CSS/:defined) pseudo-class for custom elements, e.g. https://store.google.com/us/product/pixel_7?hl=en-US
In the previous implementation, we don't define any custom elements during the playback so all :defined styles are not applied.
In this pull request, I add the support for replaying :defined pseudo-class of custom elements.

Replayer before this PR:
image

Replayer after this PR:
image

### CODE REVIEW

**Comment by Juice10** [comment]
This is great @YunFeng0817! One thing I’m wondering is if [CustomElementRegistry.define()](https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry/define) was never called during recording time this might lead to unexpected behavior. Maybe we should monkeypatch [CustomElementRegistry.define()](https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry/define) to make sure these are accurately captured

**Comment by YunFeng0817** [comment]
Good suggestion. Do you think it's OK to change the data structure SerializedElement to this
```
export type elementNode = {
  type: NodeType.Element;
  tagName: string;
  attributes: attributes;
  childNodes: serializedNodeWithId[];
  isSVG?: true;
  needBlock?: boolean;
  isCustom?: boolean; // determine whether this is a defined custom element
};
```

**Comment by Juice10** [comment]
@YunFeng0817 yes I think that works

**Comment by Juice10** [comment]
@YunFeng0817 I think I've found another edge case for you that currently isn't handled by this PR

- An element gets added to the dom eg. `special-element` (not yet `:defined` or `isCustom`
- After a couple seconds `CustomElementRegistry.define()` gets called and `special-element` is officially defined, triggering `:defined` styles

**Review by Juice10 — APPROVED** [review]
This is really great! Thanks @YunFeng0817!

**Comment by Juice10** [comment]
Looks like you have a failing type check. When that's fixed feel free to merge this

**Comment by bruno-garcia** [comment]
Hey folks, any blockers to merging this? It does help! Sentry has cherry picked this fix into our fork.

**Comment by nafees87n** [comment]
Any blockers to this PR? We really need this fix

File: packages/rrweb/src/record/observer.ts [hunk 1]
```diff
@@ -1169,6 +1170,41 @@ function initSelectionObserver(param: observerParam): listenerHandler {
...
+        constructor: CustomElementConstructor,
+        options?: ElementDefinitionOptions,
+      ) {
+        try {
+          customElementCb({
+            define: {
+              name,
+            },
+          });
+        } catch (e) {}
```
Comment (Juice10) [inline packages/rrweb/src/record/observer.ts:1200]: Fix eslint issue
```suggestion
        } catch (e) {
          console.warn(`Custom element callback failed for ${name}`);
        }
```

File: packages/rrweb/src/record/observer.ts [hunk 2]
```diff
@@ -1169,6 +1170,43 @@ function initSelectionObserver(param: observerParam): listenerHandler {
   return on('selectionchange', updateSelection);
 }
 
+function initCustomElementObserver({
+  doc,
+  customElementCb,
+}: observerParam): listenerHandler {
+  const win = doc.defaultView as IWindow;
+  if (!win || !win.customElements) return () => {};
```
Comment (Juice10) [inline packages/rrweb/src/record/observer.ts:1178]:
```suggestion
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  if (!win || !win.customElements) return () => {};
```

**Pull request merged**